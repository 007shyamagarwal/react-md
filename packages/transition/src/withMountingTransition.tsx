import * as React from "react";
import * as PropTypes from "prop-types";
import cn from "classnames";

export declare type TransitionEvent = React.TransitionEvent<HTMLElement | React.ReactHTMLElement<any>>;

export interface IMountingTransitionProps {
  /**
   * An optional className to apply.
   */
  className?: string;

  /**
   * Boolean if the component should be visible. When this prop changes from `false` to `true`,
   * the enter transition will start. When this prop changes from `false` to `true`, the leave
   * transition will start.
   */
  visible: boolean;

  /**
   * An optional function to call when the enter or leave transition ends.
   */
  onTransitionEnd?: (e: TransitionEvent) => void;
}

export interface IMountingTransitionState {
  rendered: boolean;
  transitioning: boolean;
  transitioningActive: boolean;
}

export interface IInjectedMountingTransitionProps {
  /**
   * The full component's class name. This will be updated to include the transition names
   * during the transition and return back to the empty string or provided className prop
   * once the transition has finished.
   */
  className: string;

  /**
   * Boolean if the component is currently transitioning.
   */
  transitioning: boolean;

  /**
   * A function to call when the enter or leave transition has finished. You are required to
   * apply this function to the child component so that the transitions will finish. If this
   * is excluded, the component will never unmount fully.
   */
  onTransitionEnd: (e: TransitionEvent) => void;
}

/**
 * This is like another version of CSSTransitionGroup (v1) that renders null instead
 * of a span when empty and only works for one element at a time.
 *
 * When the visible prop is false, nothing will be rendered. When the visible prop is switched
 * to true, it will start the in transition by:
 * - rendering the component as normal (so no additional class names)
 * - providing the component with an enter transition class name
 * - providing the component with an enter and active transition class name
 * - rendering the component as normal once transition has finished (so no additional class names)
 *
 * Now when the visible prop is switched to false, it will start the leave transition with the same flow as
 * above but reversed. Once the transition is done, it will render nothing again.
 *
 * The class names generated will be:
 * - `${transitionName}--enter`
 * - `${transitionName}--enter-active`
 * - `${transitionName}--leave`
 * - `${transitionName}--leave-active`
 *
 * So to get the animation working correctly, your component being wrapped with this HOC **must** apply the
 * provided className prop and apply the onTransitionEnd prop that gets passed down.
 *
 * @param {String} transitionName - The transition name to use for the enter and leave transitions. Suffixes will be applied
 *    this this string for each stage of the transition.
 * @return {function} a higher order component creator function.
 */
export default function withMountingTransition<ComponentProps extends {}>(transitionName: string) {
  return function withMountingTransitionClassName(Component: React.ComponentType<ComponentProps & IInjectedMountingTransitionProps>) {
    type WithMountingTransitionProps = ComponentProps & IMountingTransitionProps;
    let displayName = 'WithMountingTransition';
    if (process.env.NODE_ENV !== "produciton") {
      displayName = `withMountingTransition(${Component.displayName})`;
    }

    return class WithMountingTransition extends React.Component<WithMountingTransitionProps, IMountingTransitionState> {
      public static displayName = displayName;
      public static propTypes = {
        className: PropTypes.string,
        visible: PropTypes.bool.isRequired,
        onTransitionEnd: PropTypes.func,
      };

      public static getDerivedStateFromProps(nextProps: ComponentProps & IMountingTransitionProps, prevState: IMountingTransitionState) {
        if (nextProps.visible && !prevState.rendered && !prevState.transitioning) {
          return { rendered: true, transitioning: true, transitioningActive: false };
        } else if (!nextProps.visible && prevState.rendered && !prevState.transitioning) {
          return { transitioning: true, transitioningActive: false };
        }

        return null;
      }

      public frame: number | null;

      constructor(props: ComponentProps & IMountingTransitionProps) {
        super(props);

        this.frame = null;
        this.state = {
          rendered: props.visible,
          transitioning: false,
          transitioningActive: false,
        };
      }

      public componentDidUpdate(prevProps: ComponentProps & IMountingTransitionProps, prevState: IMountingTransitionState) {
        if (this.state.transitioning && !prevState.transitioning) {
          this.frame = window.requestAnimationFrame(() => {
            this.frame = null;
            this.setState({ transitioningActive: true });
          });
        }
      }

      public componentWillUnmount() {
        this.clear();
      }

      public render() {
        const { rendered, transitioning, transitioningActive } = this.state;
        if (!rendered) {
          return null;
        }

        const { className, visible, ...props } = this.props as IMountingTransitionProps;
        return (
          <Component
            {...props}
            className={cn({
              [`${transitionName}--enter`]: visible && transitioning,
              [`${transitionName}--enter-active`]: visible && transitioning && transitioningActive,
              [`${transitionName}--leave`]: !visible && transitioning,
              [`${transitionName}--leave-active`]: !visible && transitioning && transitioningActive,
            }, className)}
            transitioning={transitioning}
            onTransitionEnd={this.handleTransitionEnd}
          />
        );
      }

      public clear = () => {
        if (this.frame) {
          window.cancelAnimationFrame(this.frame);
          this.frame = null;
        }
      }

      public handleTransitionEnd = (e: TransitionEvent) => {
        if (this.props.onTransitionEnd) {
          this.props.onTransitionEnd(e);
        }

        if (this.state.transitioning || this.state.transitioningActive) {
          this.setState({
            rendered: this.props.visible,
            transitioning: false,
            transitioningActive: false,
          });
        }
      }
    }
  }
}
