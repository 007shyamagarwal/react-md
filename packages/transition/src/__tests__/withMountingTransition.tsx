import * as React from "react";
import { mount } from "enzyme";
import renderer from "react-test-renderer";
import { default as withMountingTransition, IInjectedMountingTransitionProps } from "../withMountingTransition";

describe("withMountingTransition", () => {
  const wrapper = withMountingTransition("some-class-name");

  it("should render correctly based on the visible prop", () => {
    class Test extends React.Component<IInjectedMountingTransitionProps, {}> {
      render() {
        const { transitioning, ...props } = this.props;
        return <div {...props} />;
      }
    }

    const TestWithAnimation = wrapper(Test);
    expect(renderer.create(<TestWithAnimation visible={false} />)).toMatchSnapshot();
    expect(renderer.create(<TestWithAnimation visible={true} />)).toMatchSnapshot();
  });

  it("should follow the rendering flow when the visible prop is changed from false to true", () => {
    // fake timers so we can step-by-step render snapshots during the transitions
    // also "mock" the requestAnimationFrame so that it uses a timeout since I don't know a way to
    // get jest to update after an animation frame
    jest.useFakeTimers();
    const { requestAnimationFrame } = global;
    global.requestAnimationFrame = cb => setTimeout(cb, 0);

    class Test extends React.Component<IInjectedMountingTransitionProps, {}> {
      render() {
        const { transitioning, ...props } = this.props;
        return <div {...props} />;
      }
    }

    const TestWithAnimation = wrapper(Test);

    // test enter animation
    const test = mount(<TestWithAnimation visible={false} />);
    expect(test.render()).toMatchSnapshot();

    test.setProps({ visible: true });
    expect(test.render()).toMatchSnapshot();

    jest.runOnlyPendingTimers();
    expect(test.render()).toMatchSnapshot();

    test.simulate("transitionEnd");
    expect(test.render()).toMatchSnapshot();

    // test leave animation
    test.setProps({ visible: false });
    expect(test.render()).toMatchSnapshot();

    jest.runOnlyPendingTimers();
    expect(test.render()).toMatchSnapshot();

    test.simulate("transitionEnd");
    expect(test.render()).toMatchSnapshot();

    global.requestAnimationFrame = requestAnimationFrame;
  });
});
