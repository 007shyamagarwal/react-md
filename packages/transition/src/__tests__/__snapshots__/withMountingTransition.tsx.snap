// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withMountingTransition should follow the rendering flow when the visible prop is changed from false to true 1`] = `null`;

exports[`withMountingTransition should follow the rendering flow when the visible prop is changed from false to true 2`] = `
<div
  class="some-class-name--enter"
/>
`;

exports[`withMountingTransition should follow the rendering flow when the visible prop is changed from false to true 3`] = `
<div
  class="some-class-name--enter some-class-name--enter-active"
/>
`;

exports[`withMountingTransition should follow the rendering flow when the visible prop is changed from false to true 4`] = `
<div
  class=""
/>
`;

exports[`withMountingTransition should follow the rendering flow when the visible prop is changed from false to true 5`] = `
<div
  class="some-class-name--leave"
/>
`;

exports[`withMountingTransition should follow the rendering flow when the visible prop is changed from false to true 6`] = `
<div
  class="some-class-name--leave some-class-name--leave-active"
/>
`;

exports[`withMountingTransition should follow the rendering flow when the visible prop is changed from false to true 7`] = `null`;

exports[`withMountingTransition should render correctly based on the visible prop 1`] = `null`;

exports[`withMountingTransition should render correctly based on the visible prop 2`] = `
<div
  className=""
  onTransitionEnd={[Function]}
/>
`;
